src/openrct2/peep/Guest.cpp

```diff
@@ -4880,55 +4882,131 @@ void Guest::UpdateRideMazePathfinding()
+            {
+                currentEntry.MarkVisited(sourceDirection, openEdgesCount);
+                // Mark also the previous chosen edge in case of dead-end-loop
+                if (currentEntry.MatchCoords(previousEntry))
+                    currentEntry.MarkVisited(PeepMazeRegister.DirectionAtLastIntersection, openEdgesCount);
+            }
         }
-        hedges[openCount++] = mazeReverseLastEdge;
-    }
 
+        // Probability table that the guest remember of the already taken edges
+        // TODO: Tune these values
+        static const std::array<std::array<uint16_t, 4>, 6> rememberThreshold = {
+            { // visited,               origin,                 visited+peeked,         origin+peeked
+              // Last intersection
+              { maze_forget_odds(0.04), maze_forget_odds(0.02), maze_forget_odds(0.00), maze_forget_odds(0.00) },
+              // -1
+              { maze_forget_odds(0.09), maze_forget_odds(0.03), maze_forget_odds(0.03), maze_forget_odds(0.01) },
+              // -2
+              { maze_forget_odds(0.27), maze_forget_odds(0.18), maze_forget_odds(0.07), maze_forget_odds(0.03) },
+              // -3
+              { maze_forget_odds(0.39), maze_forget_odds(0.25), maze_forget_odds(0.15), maze_forget_odds(0.10) },
+              // -4
+              { maze_forget_odds(0.50), maze_forget_odds(0.45), maze_forget_odds(0.32), maze_forget_odds(0.30) },
+              // New or forgotten intersection
+              { maze_forget_odds(1.00), maze_forget_odds(1.00), maze_forget_odds(1.00), maze_forget_odds(1.00) } }
+        };
+
+        // If the guest succeeded his memory roll, remove visited edges (other that the origin edge)
+        if ((scenario_rand() & 0xffff) >= rememberThreshold[entryIndex][peekFlag * 2])
+            openEdgesFlags &= ~currentEntry.GetVisited();
+
+        // Second dice roll, if succeeded remove also the origin edge (unless all the other edges are already visited)
+        if (bitcount(openEdgesFlags) > 1 && (scenario_rand() & 0xffff) >= rememberThreshold[entryIndex][1 + peekFlag * 2])
+            openEdgesFlags &= ~(0x1 << currentEntry.GetOrigin());
+    }
+    if (openEdgesFlags == 0)
+        openEdgesFlags |= (1 << sourceDirection);
+
+    // Selection randomly an available direction
+    uint8_t hedges[4]{ 0xFF, 0xFF, 0xFF, 0xFF };
+    uint8_t openCount = bitcount(openEdgesFlags);
+    for (int i = 0, j = 0; i < 4 && j < openCount; ++i)
+    {
+        if (openEdgesFlags & (1 << i))
+            hedges[j++] = i;
+    }
     uint8_t chosenEdge = hedges[scenario_rand() % openCount];
     assert(chosenEdge != 0xFF);
 
+    if (openEdgesCount > 2)
+        PeepMazeRegister.DirectionAtLastIntersection = chosenEdge;
+
     targetLoc.x = CoordsDirectionDelta[chosenEdge].x / 2;
     targetLoc.y = CoordsDirectionDelta[chosenEdge].y / 2;
 
     targetLoc.x += DestinationX;
     targetLoc.y += DestinationY;
 
-    enum class maze_type
+    enum class maze_next_tile_type
```

Comment:
This enum is used locally (so that why I permitted me this modification) and I changed its name to reflect better that the code searches to characterize the type of the next tile and not the type of a maze in its entirety.